ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -125,6 +127,32 @@ public int compare(IndexEntry e1, IndexEntry e2) {
   private final IndexPersistor persistor = new IndexPersistor();
   private final ScheduledFuture<?> persistorTask;
 
+  // ReadWriteLock to protect index values are always pointing to a valid log segments
+  // In compaction's final steps, persistent index and log will update their internal maps to reflect the result of
+  // compactions. This might have race condition with getBlobReadInfo method
+  // --------------------------------------------------------------------------------------
+  // compaction thread                          | getBlobReadInfo thread
+  // --------------------------------------------------------------------------------------
+  // 1. RenameTempLogSegmentFilenames           | 1. FindKeyToReturnIndexValue
+  // 2. AddNewLogSegmentToLogMap                |
+  // 3. UpdateIndexMap                          |
+  // 4. WaitForLogSegmentRefCountToZero         |
+  // 5. RemoveIndexSegmentFiles                 |
+  // 6. RemoveOldLogSegmentFromLogMap           |
+  // 7. RemoveLogSegmentFiles                   | 2. GetLogSegmentFromIndexValueFromLogMap
+  //                                            | 3. IncreaseLogSegmentRefCount
+  // -------------------------------------------------------------------------------------
+  // If two threads are executing in the above given order, then the GetLogSegmentFromIndexValueFromLogMap would
+  // throw an NullPointerException since RemoveOldLogSegmentFromLogMap would remove it from the map.
+  // This ReadWriteLock would prevent that from happening.
+  // 1. In getBlobReadInfo method, we use read lock to lock all three steps
+  // 2. In compaction thread, we use write lock to lock step 3 (in method changeIndexSegments)
+  // If getBlobReadInfo happens before compaction's step 3, then compaction thread would block at acquiring
+  // the write lock, so getBlobReadInfo's step 2 would not happen after compaction's step6.
+  // If getBlobReadInfo happens after compaction's step 3, then the IndexValue returned by findKey method would reflect
+  // the result of the compaction already, it will not point to any old log segment.
+  private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
```

Comment:
The scope of this lock is not correct. This lock needs to be owned by Blobstore and passed to this class. See the comment below.

Also, using a single lock seems to be too coarse. Do you want to block all reads, etc when changeIndexSegments run or prefer blocking access to the index/log segments only impacted by the compaction? I think creating a lock per index, log-segment tuple makes more sense.

Reply:
Commented below, the scope should be sufficient to guarantee the correctness.

Using a single lock should be fine. The write lock would only be used in compaction, which happens either once in 24 hours or once in one hour. And the read lock's performance shouldn't be an issue.